### DISCUSSION TITLE
M106 P1 - P3 mix up for Creality K1

### DISCUSSION THREAD

**Discussion by DSPrints** [OrcaSlicer/OrcaSlicer#3878] [Dev]
Orca Slicer has the exhaust fan feature that doesn't work for how a K1 is set up. 

<img width="368" alt="Screenshot 2024-01-28 at 5 18 35 pm" src="https://github.com/SoftFever/OrcaSlicer/assets/155401306/22b9e268-a37d-4139-9778-19c0cae66c58">

This Option in the slicer activates M106 P3 fan, which isn't a fan on the K1. 
M106 P0 = Model/extruder cooling fan
M106 P1 = Chamber Exhaust fan
M106 P2 = Side Cooling fan

If I edit the GCODE in text edit, swapping P3 for P1 everywhere it appears, then it works perfectly. 👍
Can this be amended? Is there a work around other then turning it off & manually inserting M106 P1 S255 to the filament GCODE?

Additionally, the completed print %, it would be good if it could be set for a timer duration? So the fan stays on for 2 minutes after print, then turns off.. not sure how to implement that in a GCODE.

**Comment by bnc-rainman** [comment]
If the prophet does not come to the mountain, the mountain must come to the prophet.

https://guilouz.github.io/Creality-Helper-Script-Wiki/helper-script/fans-control-macros/

This helper scripts activates M106 P3 for chamber fan control and add M141 and M191 for chamber temperatur control

**Comment by onkytonk** [comment]
I've run into the same issue now trying to setup for ABS/ASA. 

Did you ever find a solution apart from per filament gcode?

**Comment by mechano** [comment]
But it's not an issue of Orca. 

You can change the behaviour of everything from macros.
Just take inspiration from the following code from TwoTrees SK1. 
In the macro change the fan name with the fan you prefer when Orca calls M106 P{x} S{x}. 

```
# Fan pin definitions
[fan]
pin: MKS_THR:gpio3

[fan_generic auxiliary_fan]
pin: PC9
max_power: 1.0
shutdown_speed: 0.0
cycle_time: 0.0005 #2 kHz PWM signal
hardware_pwm: False
kick_start_time: 0

[fan_generic filter_fan]
pin: PC12
max_power: 1.0
shutdown_speed: 0.0
cycle_time: 0.0005 #2 kHz PWM signal
hardware_pwm: False
kick_start_time: 0

# Used by Orca Slicer to control also Auxiliary and Filter fan
[gcode_macro M106]
rename_existing: M106.1
gcode:
  {% set fan = 'fan' + (params.P|int if params.P is defined else 0)|string %}
  {% set speed = (params.S|float / 255 if params.S is defined else 1.0) %}
  {% if fan == 'fan2' %}
    SET_FAN_SPEED FAN=auxiliary_fan SPEED={speed}
  {% endif %}
  {% if fan == 'fan3' %}
    SET_FAN_SPEED FAN=filter_fan SPEED={speed}
  {% endif %}
  {% if params.S is defined and fan == 'fan0' %}
    {% set speed = params.S|string %}
    M106.1 S{speed}
  {% endif %}
```

**Comment by micaheli** [comment]
This is still very much an issue as of 2.3.0-beta - build dc49116  -- In my instance, I'm not using klipper. I'm using duet/RRF.  Even with exhaust fan and auxillary fan options unchecked/disabled (or set to 0), fan 3 is invoked. I do not have a fan 3, so every print I get an error saying that it cannot turn on F3 as it does not exist.  Why is gcode being generated for a disabled option and why is there no option to assign a fan number to each fan function?

**Comment by juanemartinboix** [comment]
Same here

**Comment by webercruz** [comment]
In the 'gcode_macro.cfg' file, look for the '[gcode_macro M106]' macro.
Right below the line '{% set tmp = params.P|int %}', add these three lines:
{% if tmp == 3 %}
  {% set tmp = 1 %}
{% endif %}
That's it; if the G-code contains M106 P3 S---, it will execute M106 P1 S--- and the exhaust fan will operate normally.
If the G-code contains M106 P1 S---, it will also work normally.
